### ISSUE TITLE
Improve `StringBuf.add` to increase performance

### ISSUE THREAD

**Issue by back2dos** [HaxeFoundation/haxe#5440]
My problem lies with this:

```haxe
class StringBuf {
	public inline function add<T>(x :T):Void;
}
```

Because of this signature, anything added to a `StringBuf` gets wrapped in `Std.string`, which is true for `String` itself.

To milk the irony of the situation: there is no fast way to add a `String` to a `StringBuf`. 

Therefore I propose this change:

```haxe
class StringBuf {
	public inline function add(x:StringBufFragment):Void;
}

private abstract StringBufFragment(String) from String to String {
  // possibly add fast paths here for @:from Int/Float/Bool
  @:from static inline function ofAny<A>(a:A):StringBufFragment return Std.string(a);
}
```

That gives me a speedup of 1.5 - 3.5 depending on browser: http://try.haxe.org/#b3b92

What's more, we can see there's far less code in the output too: http://try.haxe.org/#F5816

Would such a change be ok?

**Comment by jdonaldson** [comment]
I thought the whole point of StringBuf was to store the added string
fragments in an array and only add them together when toString() was
called.  Am I missing something?

**Comment by kevinresol** [comment]
The actual implementations are different per platform, some use a string, some use array or others.

**Comment by back2dos** [comment]
@jdonaldson that's not really my point here. My issue is with the fact that there is no way to directly add a `String` to a `StringBuf` and that's rather silly, to put it bluntly. The problem is not in the implementation, it is in the signature. Either we add an `addString` method or we use the abstract I proposed.

With the abstract, existing code will be sped up by a significant factor. I realized though that there's a downside also, namely the risk of breaking code that relies on `StringBuf.add` unifying with any `Whatever->Void`. I think that's worth it. Otherwise we still have a more conservative option.

**Comment by jdonaldson** [comment]
Wow, an array based stringbuf is faster than the current one, but Juraj's
is still faster.  Interesting.  Nice find Juraj.

http://try.haxe.org/#3013d

**Comment by kevinresol** [comment]
> Wow, an array based stringbuf is faster than the current one

@jdonaldson probably not. You are adding a string directly, that skipped the overhead of `Std.string` of the original `b+=x` where x is generic T which gets wrapped by `Std.string`

**Comment by jdonaldson** [comment]
I understand.  Yeah, I think just an extra addString method would make
sense.

**Comment by jdonaldson** [comment]
One more question then, why not make the entire StringBuf an abstract?
http://try.haxe.org/#0A4B1

**Comment by nadako** [comment]
> One more question then, why not make the entire StringBuf an abstract?

That question applies to many Haxe std classes (e.g. Bytes, Map impls and whatnot) :)
But since it's a breaking change and we still have Dynamic, we can't change that.

**Comment by back2dos** [comment]
>  One more question then, why not make the entire StringBuf an abstract?

If you do that you get value semantics rather than reference semantics. To clarify that point, let's imagine a function like this:

```haxe
function doFoo(output:StringBuf):Void {
  //add some stuff to the output
}
```

OTOH the analyzer is able to not allocate the wrapper much of the time.

**Comment by jdonaldson** [comment]
good point

**Comment by sebthom** [comment]
I am getting very mixed results when executing that micro benchmark, esp. when changing the test order. Probably some JIT optimizations kicks in at some point.

In Neko 2.1 + Haxe 3.30rc1 I am getting results that do not look good at all:
StringBuf: 2s
ArrayStringBuf: 4s
FastBuf: aborted after 1min with 100% cpu usage

**Comment by sebthom** [comment]
C#, Java show the same behavior as Neko. StringBuf is the fastest, them comes ArrayStringBuf and FastBuf doesn't seem to finish at all.

In C++ ArrayStringBuf is fastes, then comes StringBuf and FastBuf is 20 times slower than StringBuf.

In PHP FastBuf is MUCH faster than StringBuf or ArrayStringBuf.

Python and Flash show similar behavior to JS.

So before considering doing the proposed change please test all other targets as well.

It does not seem to be usable on Neko, C#, Java and C++.

**Comment by sebthom** [comment]
I did some more tests solely comparing the performance of a `Array<String>` vs `StringBuf`. 
`StringBuf` is considerably faster on Java, Neko and C# whereas using `Array<String>` for
string concatenation is much faster on JavaScript, Fash, PHP, Python and C++.

**Comment by kevinresol** [comment]
@sebthom can you post your test code?

**Comment by sebthom** [comment]
I am just using Justin's test case. http://try.haxe.org/#0A4B1

**Comment by kevinresol** [comment]
> I did some more tests solely comparing the performance of a Array<String> vs StringBuf.

I mean this one?

**Comment by kevinresol** [comment]
@sebthom if guess you are talking about this one http://try.haxe.org/#3013d
As I said, it is not apple-to-apple comparison, the signature of `add(s:String)` saved you a `Std.string()` call compared to `add<T>(s:T)`

**Comment by sebthom** [comment]
Right, then the Std.string overhead on some targets seems to be really huge.

**Comment by sebthom** [comment]
You are right, changing the `ArrayStringBuf` to mimic the `StringBuf` signature makes it slower compared to `StringBuf` and `Array<String>` on all platforms:
```haxe
class ArrayStringBufGeneric{
	var b = [];
    public inline function new(){}
    public inline function add<T>(x:T):Void
        b.push(Std.string(x));
    public inline function toString() return b.join("");
}
```
To sum it up, `Std.string` really impacts performance but the proposed solution using an abstract improves it on some platforms but massively slows down `StringBuf` on Neko, C#, Java and C++.

**Comment by sebthom** [comment]
After digging into the generated code of the test example on the java platform. I realized that the micro benchmark cannot be applied as-is to Java, because the `FastBuf` uses += string concatenation which produces loads of temp objects. 
If you change `FastBuf` to use a string array but keep the `StringBufFragment` as parameter type in the method signature it works fine on all platforms. Then you have best of both worlds, the performance of `ArrayStringBuf` and a method that accepts any type.

**Comment by back2dos** [comment]
I'm sorry if I was unclear. 

I am in no way suggesting to change the implementation, so everybody, let's please stop discussing it here. If there is a case to be made for some implementation yielding better performance on some targets, then that's great - go ahead and open an issue about it.

My point is that the current *signature* causes the output to be littered with undifferentiated `Std.string` calls, resulting in far worse performance and rather significant increase in output size (at least for haxe/js).

**Comment by sebthom** [comment]
I think I created the confusion as I simply ran the provided benchmark on all platforms and saw that it went rampage on some of them. The performance degradation was not because of your suggested change but because of the test setup. I now believe that your change is safe to be applied to all platforms.

**Comment by jdonaldson** [comment]
My 2 cents is to add the addString method, unless you think the abstract string input argument really is a big win.

On v2 of the stdlib, there's a bunch of places we could be using abstracts for a cleaner/faster implementation.  I think an abstract for stringbuf would fit in there (one that preserves the reference semantics).

**Comment by back2dos** [comment]
> My 2 cents is to add the addString method, unless you think the abstract string input argument really is a big win.

A really big win would be to go back in time and have `add : String->Void` and have an `addValue<T> : T-> Void` for convenience :D

But were stuck here, 10 years after that design decision has been made. In that position, the abstract input argument is an attempt to strike a compromise: get *almost* full backward compatibility and significantly improve the generated result for already existing code. This does affect `haxe.Template`, `haxe.Serializer`, `haxe.xml.Printer`, `haxe.format.JsonPrinter` and what not. Which makes me think that breaking the rather exotic usage I pointed out is well worth the benefits.

**Comment by nadako** [comment]
Sounds reasonable to me :)

**Comment by sebthom** [comment]
I too believe that the change as proposed is desirable because all existing code will instantly benefit from the performance improvement without any code changes.

**Comment by jdonaldson** [comment]
Let me throw out one more idea, just to play devil's advocate.   What if we add special abstract behavior for Std.string argument instead?   Wouldn't that cover more territory?

**Comment by ncannasse** [comment]
I don't think we want to overcomplexify the StringBuf.add for that.
We could have addString as a specialized/optimized method for String, that seems more reasonable.

**Comment by Simn** [comment]
What's our policy regarding `null` here? If we still have to branch in order to detect it we'd not really gain anything.

**Comment by ncannasse** [comment]
Yes of course, or on some platforms we might simply put it as-it (Array based impl for instance)

**Comment by Simn** [comment]
Ah, I thought it was necessary to check for `null` anyway so `Std.string(string)` wouldn't really be better or worse. But the argument makes sense to me if we can skip that check.

However, I want to mention that having to mangle type names into method names strikes me as somewhat weak in a rich typing system like we have. Shouldn't we have a way to detect that we're dealing with `add(string)` and do whatever we would do in `addString` automatically?

**Comment by ncannasse** [comment]
@Simn I'm interested if we find an elegant way to add that to the type system, but I don't have any illuminating idea for it atm.

Juraj solution works we would like a more general and less verbose solution that can also display nicely in the documentation. For the reminder we had a similar issue with Lambda methods which we would like to accept both Iterator and Iterable

We could tag methods with a specific metadata : `@:variant(addString) function add( v : Dynamic )` but that's not very documentation-friendly.

Another way is to have some new specific compile-time type-checking operator that gets eliminated before generation:

```
inline function add( v : Dynamic ) {
    if( v isRuntime String ) addString(v) else addDynamic(v)
}
```

Then a more general approach is to have union types + compile time optimization of type switches:

```
inline function add( v : String | Dynamic ) {
    switch( v ) {
    case String: addString(v);
    case Dynamic: addDynamic(v);
    }
}
```

I wonder how hard the later would be add to our type system.

**Comment by Simn** [comment]
I'm very interested in these union types and already thought about it a bit. From an implementation point of view it's quite similar to GADT with per-case specialization. I think I could prototype this fairly quickly, maybe using EitherType as a demo.

**Comment by RealyUniqueName** [comment]
I think overloading was invented for cases just like this ;)

**Comment by jdonaldson** [comment]
addString in 3.4, abstract in 4.0?

**Comment by frabbit** [comment]
can we do something here? The unicode implementation would profit from having this.

**Comment by Simn** [comment]
My preferred solution would be to support `@:overload inline function` on all targets and force the inlining.

**Comment by frabbit** [comment]
would be nice to find a solution for this issue.
